### PR TITLE
Generics in core

### PR DESCRIPTION
**Pull request by twogee**
This is as far as I can get without significant changes to API.
Any remaining points of contention are annotated by `@SuppressWarnings` (and I'm not proud of what `NoFilter` looks like).
If you need proof that arrays of generics are broken, just change `List<T>` back to `T[]` in `ModuleRules` and watch the fireworks.

### CODE REVIEW

File: src/java/org/apache/ivy/osgi/p2/XMLInputParser.java [hunk 1]
```diff
@@ -19,12 +19,11 @@
...
 import java.io.IOException;
 import java.io.InputStream;
-import java.text.ParseException;
 
 import org.xml.sax.SAXException;
 
 public interface XMLInputParser {
 
-    void parse(InputStream in) throws ParseException, IOException, SAXException;
+    void parse(InputStream in) throws IOException, SAXException;
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/p2/XMLInputParser.java]: Given that `ParseException` is a checked exception, removing this from the throws clause can cause existing callers of this method (potentially outside of Ivy) to run into compile issues. Is there some specific reason why this is being removed from the throws clause? If not, can you please undo this change.

File: src/java/org/apache/ivy/osgi/repo/AbstractFSManifestIterable.java [hunk 2]
```diff
@@ -44,9 +44,9 @@ public AbstractFSManifestIterable(T root) {
         return new FSManifestIterator();
     }
 
-    protected abstract List<T> listBundleFiles(T dir) throws IOException;
+    protected abstract List<T> listBundleFiles(T dir);
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/repo/AbstractFSManifestIterable.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/osgi/repo/AbstractFSManifestIterable.java [hunk 3]
```diff
@@ -44,9 +44,9 @@ public AbstractFSManifestIterable(T root) {
...
 
-    protected abstract List<T> listDirs(T dir) throws IOException;
+    protected abstract List<T> listDirs(T dir);
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/repo/AbstractFSManifestIterable.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/osgi/repo/RepositoryManifestIterable.java [hunk 4]
```diff
@@ -59,11 +59,11 @@ protected InputStream getInputStream(String f) throws IOException {
         return repo.getResource(f).openStream();
     }
 
-    protected List<String> listBundleFiles(String dir) throws IOException {
+    protected List<String> listBundleFiles(String dir) {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/repo/RepositoryManifestIterable.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/osgi/repo/RepositoryManifestIterable.java [hunk 5]
```diff
@@ -59,11 +59,11 @@ protected InputStream getInputStream(String f) throws IOException {
...
         return asList(ResolverHelper.listAll(repo, dir));
     }
 
-    protected List<String> listDirs(String dir) throws IOException {
+    protected List<String> listDirs(String dir) {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/repo/RepositoryManifestIterable.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/osgi/util/VersionRange.java [hunk 6]
```diff
@@ -306,7 +306,7 @@ public boolean isClosedRange() {
         return startVersion.equals(endVersion);
     }
 
-    public boolean contains(String versionStr) throws ParseException {
+    public boolean contains(String versionStr) {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/osgi/util/VersionRange.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/plugins/repository/sftp/SFTPRepository.java [hunk 7]
```diff
@@ -148,7 +142,7 @@ public void put(File source, String destination, boolean overwrite) throws IOExc
         }
     }
 
-    private void mkdirs(String directory, ChannelSftp c) throws IOException, SftpException {
+    private void mkdirs(String directory, ChannelSftp c) throws SftpException {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/plugins/repository/sftp/SFTPRepository.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/plugins/repository/ssh/SshCache.java [hunk 8]
```diff
@@ -263,9 +263,8 @@ public void clearSession(Session session) {
      * @param session
      *            to connect to
      * @return channelSftp or null if not successful (channel not existent or dead)
-     * @throws IOException if something goes wrong
      */
-    public ChannelSftp getChannelSftp(Session session) throws IOException {
+    public ChannelSftp getChannelSftp(Session session) {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/plugins/repository/ssh/SshCache.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

File: src/java/org/apache/ivy/tools/analyser/JarJarDependencyAnalyser.java [hunk 9]
```diff
@@ -39,21 +39,22 @@ public JarJarDependencyAnalyser(File jarjarjarLocation) {
...
-                jarjarCmd.append(File.pathSeparator);
-            }
+                    jarModule.getMrid(), new Date(jarModule.getJar().lastModified()));
+            mds.put(jarModule.getMrid(), md);
+            jarjarCmd.append("\"").append(jarModule.getJar().getAbsolutePath()).append("\"");
+            jarjarCmd.append(File.pathSeparator);
+        }
+
+        if (modules.length > 0) {
+            jarjarCmd.setLength(jarjarCmd.length() - 1);
```
Comment (jaikiran) [inline src/java/org/apache/ivy/tools/analyser/JarJarDependencyAnalyser.java]: I don't fully follow what this is block is meant for. Is it to conditionally include the `File.pathSeparator` character?

File: src/java/org/apache/ivy/util/url/HttpClientHandler.java [hunk 10]
```diff
@@ -177,7 +177,7 @@ public URLInfo getURLInfo(URL url, int timeout) {
         return UNAVAILABLE;
     }
 
-    private boolean checkStatusCode(URL url, HttpMethodBase method) throws IOException {
+    private boolean checkStatusCode(URL url, HttpMethodBase method) {
```
Comment (jaikiran) [inline src/java/org/apache/ivy/util/url/HttpClientHandler.java]: Similar comment as here https://github.com/apache/ant-ivy/pull/52/files#r127395633 applies to this line

**Comment by jaikiran** [comment]
I've reviewed this PR and apart from the review comments noted, the rest of the changes look fine. Please only include any review comments related changes to this PR, from now on. Else it's going to take a massive effort to review this PR afresh again. Also, while doing any new review comments related changes to this specific PR, please do them in new commits - that way, I'll know existing commits haven't changed. While merging upstream, I'll make sure that I squash those commits myself.

File: src/java/org/apache/ivy/tools/analyser/JarJarDependencyAnalyser.java (hunk 9 above)
Comment (twogee) [inline src/java/org/apache/ivy/tools/analyser/JarJarDependencyAnalyser.java]: It strips an extra pathSeparator character, but only if modules has any entries. It is an afterthought on my part...

**Comment by twogee** [comment]
Thanks! I will revert the changes regarding the checked exceptions as suggested in a separate commit.

File: src/java/org/apache/ivy/plugins/repository/sftp/SFTPRepository.java (hunk 7 above)
Comment (twogee) [inline src/java/org/apache/ivy/plugins/repository/sftp/SFTPRepository.java]: This is a private method...

Comment (jaikiran) [inline src/java/org/apache/ivy/plugins/repository/sftp/SFTPRepository.java]: Agree. This change can stay. I missed this while reviewing.

File: src/java/org/apache/ivy/osgi/util/VersionRange.java (hunk 6 above)
Comment (twogee) [inline src/java/org/apache/ivy/osgi/util/VersionRange.java]: Not sure about this one, either: it's a consequence of previous changes to Version which no longer throws exceptions.

**Comment by twogee** [comment]
I committed proposed changes w/o contains(). Please let me know if should amend.

**Comment by jaikiran** [comment]
It looks like this PR has been updated with additional unrelated changes. Except for these 2 commits in this PR:
https://github.com/apache/ant-ivy/pull/52/commits/52b99a74d5770485d95d087aac8a97f8ad6c6795
https://github.com/apache/ant-ivy/pull/52/commits/f825fc31890033ed14580eb0cf06df860888bbb3

the rest of the commits related to bringing in generics to core has been merged to upstream.

These 2 commits are pretty much cosmetic changes (to large number of files) which I personally would like to avoid since they don't really bring in much value. Either way, this PR has now been closed since the major changes related to bringing in generics has now been merged.
